File: src/led_tower.cpp
#include <wiringPi.h>
#include <wiringPiI2C.h>
#include <iostream>
#include <unistd.h>

// I2C Configuration
#define I2C_BUS 1  // Normalmente 1 no Raspberry Pi
#define MCP23017_ADDRESS_1 0x24  // Endereço do MCP23017
#define MCP23017_ADDRESS_2 0x26
#define MCP23017_ADDRESS_3 0x27 


// MCP23017 Registers
#define IODIRA 0x00  // Registro de direção do Port A
#define IODIRB 0x01  // Registro de direção do Port B
#define GPIOA 0x12   // Registro GPIO Port A
#define GPIOB 0x13   // Registro GPIO Port B
#define OLATA 0x14   // Registro de latch de saída Port A
#define OLATB 0x15   // Registro de latch de saída Port B

// GPIO Pins para 74HC164
#define DATA_PIN 17
#define CLOCK_PIN 4

// Timing
#define CLOCK_DELAY 100000  // 100ms em microssegundos

// Função para configurar o MCP23017
void setupExpander(int fd) {
    // Configurar todos os pinos dos Port A e Port B como saída
    wiringPiI2CWriteReg8(fd, IODIRA, 0x00);  // Todos os pinos do Port A como saída
    wiringPiI2CWriteReg8(fd, IODIRB, 0x00);  // Todos os pinos do Port B como saída
}

// Função para configurar GPIO do MCP23017
void setGPIO_expander(int fd, char port, uint8_t value) {
    if (port == 'A') {
        wiringPiI2CWriteReg8(fd, OLATA, value);
    } else if (port == 'B') {
        wiringPiI2CWriteReg8(fd, OLATB, value);
    } else {
        std::cerr << "Porta inválida. Use 'A' ou 'B'.\n";
    }
}

// Configuração inicial do GPIO do 74HC164
void setupGPIO_shifter() {
    wiringPiSetupGpio();  // Usa o esquema BCM
    pinMode(DATA_PIN, OUTPUT);
    pinMode(CLOCK_PIN, OUTPUT);

    digitalWrite(DATA_PIN, LOW);
    digitalWrite(CLOCK_PIN, LOW);
}

// Gerar um único pulso de clock
void clockPulse() {
    digitalWrite(CLOCK_PIN, HIGH);
    usleep(CLOCK_DELAY / 2);
    digitalWrite(CLOCK_PIN, LOW);
    usleep(CLOCK_DELAY / 2);
}

// Ciclo de shift register
void shiftRegisterCycle() {
    for (int clockCount = 0; clockCount < 16; ++clockCount) {
        int init = clockCount % 2;  // 0 para par, 1 para ímpar
        digitalWrite(DATA_PIN, init);
        clockPulse();
    }
}

int main() {
    // Inicializar o barramento I2C
    int fd_1 = wiringPiI2CSetup(MCP23017_ADDRESS_1);
    if (fd_1 == -1) {
        std::cerr << "Erro ao inicializar I2C.\n";
        return -1;
    }

    int fd_2 = wiringPiI2CSetup(MCP23017_ADDRESS_2);
    if (fd_2 == -1) {
        std::cerr << "Erro ao inicializar I2C.\n";
        return -1;
    }

    int fd_3 = wiringPiI2CSetup(MCP23017_ADDRESS_3);
    if (fd_3 == -1) {
        std::cerr << "Erro ao inicializar I2C.\n";
        return -1;
    }

    // Configurar o MCP23017
    setupExpander(fd_1);
    setupExpander(fd_2);
    setupExpander(fd_3);

    setGPIO_expander(fd_1, 'A', 0xFF);  // Configurar todos os pinos do Port A como HIGH
    setGPIO_expander(fd_1, 'B', 0x00);  // Configurar todos os pinos do Port A como HIGH
    
    setGPIO_expander(fd_2, 'A', 0x00);  // Configurar todos os pinos do Port A como HIGH
    setGPIO_expander(fd_2, 'B', 0x00);  // Configurar todos os pinos do Port A como HIGH
    
    setGPIO_expander(fd_3, 'A', 0x00);  // Configurar todos os pinos do Port A como HIGH
    setGPIO_expander(fd_3, 'B', 0x00);  // Configurar todos os pinos do Port B como LOW




    try {
        setupGPIO_shifter();
        //while (true) {
        std::cout << "Iniciando um novo ciclo...\n";
        //shiftRegisterCycle();
        sleep(1);  // Espera 1 segundo antes do próximo ciclo
        //}

        
        uint8_t value = 1 << 2;  // Bit shift para criar um LED aceso em movimento
        setGPIO_expander(fd_1, 'A', value); //2, 3, 4, 5, 6, 7
        setGPIO_expander(fd_1, 'B', value); // 0, 1, 2, 3, 4, 5

        setGPIO_expander(fd_2, 'A', value);
        setGPIO_expander(fd_2, 'B', value);

        setGPIO_expander(fd_3, 'A', value);
        setGPIO_expander(fd_3, 'B', value);

        
        printf("test\n");
        usleep(1000000);  // 200ms

    } catch (...) {
        std::cout << "Saindo do programa.\n";
        digitalWrite(DATA_PIN, LOW);
        digitalWrite(CLOCK_PIN, LOW);
    }

    // Exemplo: Criar efeito "chasing" no Port A
    //for (int i = 0; i < 8; ++i) {
    //uint8_t value = 1 << 0;  // Bit shift para criar um LED aceso em movimento
    //setGPIO_expander(fd, 'A', value);
    //printf("test\n");
    //usleep(1000000);  // 200ms
    //}

    return 0;
}

File: Python/led_tower.py
import RPi.GPIO as GPIO
import smbus2
import time

# I2C Configuration
I2C_BUS = 1  # Typically 1 on Raspberry Pi
MCP23017_ADDRESS = 0x24  # 24, 26, 27

# MCP23017 Registers
IODIRA = 0x00  # I/O Direction Register for Port A
IODIRB = 0x01  # I/O Direction Register for Port B
GPIOA = 0x12   # GPIO Port A Register
GPIOB = 0x13   # GPIO Port B Register
OLATA = 0x14   # Output Latch Register for Port A
OLATB = 0x15   # Output Latch Register for Port B

# GPIO Pins for 74HC164
DATA_PIN = 17
CLOCK_PIN = 4

#Timing
CLOCK_DELAY = 0.1


def setup_expander(bus):
    """
    Initialize the MCP23017 expander.
    Set all pins on Port A and Port B as outputs.
    """
    # Set all pins of Port A and Port B as outputs (0x00 means all pins are outputs)
    bus.write_byte_data(MCP23017_ADDRESS, IODIRA, 0x00)  # Set Port A as output
    bus.write_byte_data(MCP23017_ADDRESS, IODIRB, 0x00)  # Set Port B as output

def set_gpio(bus, port, value):
    """
    Set GPIO pins on Port A or Port B.
    :param bus: SMBus object.
    :param port: 'A' for Port A or 'B' for Port B.
    :param value: 8-bit value to write to the port (e.g., 0xFF turns all pins on).
    """
    if port == 'A':
        bus.write_byte_data(MCP23017_ADDRESS, OLATA, value)
    elif port == 'B':
        bus.write_byte_data(MCP23017_ADDRESS, OLATB, value)
    else:
        raise ValueError("Port must be 'A' or 'B'.")
    
def setup_gpio_pins():
    """
    Setup the GPIO pins for the 74HC164 shift register.
    """
    GPIO.setwarnings(False)
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(DATA_PIN, GPIO.OUT)
    GPIO.setup(CLOCK_PIN, GPIO.OUT)
    
    GPIO.output(DATA_PIN, GPIO.LOW)
    GPIO.output(CLOCK_PIN, GPIO.LOW)


def clock_pulse():
    """
    Generate a single clock pulse.
    """
    GPIO.output(CLOCK_PIN, GPIO.HIGH)
    time.sleep(CLOCK_DELAY / 2)
    GPIO.output(CLOCK_PIN, GPIO.LOW)
    time.sleep(CLOCK_DELAY / 2)

def shift_register_cycle():
    """
    Perform a cycle of 16 clock pulses, alternating 'init' value:
    - Even clock: Data (init) = 0
    - Odd clock: Data (init) = 1
    """
    for clock_count in range(16):
        init = clock_count % 2  # 0 for even, 1 for odd
        GPIO.output(DATA_PIN, init)  # Set data pin
        clock_pulse()  # Generate clock pulse


def main():
    # Initialize the I2C bus
    bus = smbus2.SMBus(I2C_BUS)
    
    # Setup the MCP23017
    setup_expander(bus)
    
    # Example: Turn on all LEDs connected to Port A
    set_gpio(bus, 'A', 0xFF)  # Set all pins on Port A to HIGH
    # Example: Turn off all LEDs connected to Port B
    set_gpio(bus, 'B', 0x00)  # Set all pins on Port B to LOW
    
    try:
        setup_gpio_pins()
        while True:
            print("Starting a new cycle...")
            shift_register_cycle()
            time.sleep(1)  # Wait before starting the next cycle
    except KeyboardInterrupt:
        print("Exiting program.")
    finally:
        GPIO.cleanup()  # Reset GPIO pins on exit
    
    # Example: Create a chasing effect on Port A LEDs
    
    for i in range(8):
        value = 1 << i  # Shift a bit to create a moving "ON" LED
        set_gpio(bus, 'A', value)
        time.sleep(0.2)
    
    # Cleanup
    bus.close()

if __name__ == "__main__":
    main()
